### ISSUE TITLE
Provide a way to use a custom IWorkItemDispatcher

### ISSUE THREAD

**Issue by cwensley** [nunit/nunit#3013]
Currently there is only three options for dispatchers when using `NUnitTestAssemblyRunner`:

1. MainThreadWorkItemDispatcher
2. ParallelWorkItemDispatcher
3. SimpleWorkItemDispatcher

For UI-based applications, it would be nice to be able to invoke the tests in the UI thread, while running the unit tests in a background thread.  Currently, the only solution is to do an Invoke in each test itself, which is onerous (and messy to catch assertion exceptions and re-throw them after the invoke).

Alternatively, it would be nice to have an attribute that could specify that a test (or test fixture) should be run using `SynchronizationContext.Send()` from the `SynchronizationContext.Current` of the UI thread.

**Comment by CharliePoole** [comment]
The dispatcher is global and used for the entire run so it's not suitable to replace for a single test or fixture.

@jnm2 Didn't you do some work on this problem?

**Comment by cwensley** [comment]
@CharliePoole ah good to know, thanks!  I was thinking that I would be able to implement said attribute using a custom dispatcher, but perhaps that isn't the right place.  

Any hints of how/where something like that might be implemented would be very helpful.

**Comment by jnm2** [comment]
This seems related to https://github.com/nunit/nunit/issues/2917 (which I need to get to) and https://github.com/nunit/nunit/issues/1200 (which is about to be released in 3.11).

@cwensley Would `[Apartment(STA)]` solve your problem when you use 3.11 with https://github.com/nunit/nunit/issues/1200? I typically do UI right in the test thread, sometimes in parallel.

**Comment by cwensley** [comment]
@jnm2, Hm, with #1200 it appears I would need to return `Task` and use async/await?  I have no awaitable code, and it appears that the test itself is still initially called in the non-UI thread.

Also, it appears `[Apartment(ApartmentState.STA)]` isn't supported in netstandard, so I can't use it anyway.

#2917 is definitely what I'm after.

Thanks for the quick responses, guys.

**Comment by cwensley** [comment]
Hmm.  Just poking around more, and it looks like creating an attribute that implements `IWrapTestMethod` can potentially be used for this scenario.   I'll test it out and report back.

**Comment by cwensley** [comment]
Ok, it works!  I've used this code:

```c#
[System.AttributeUsage(AttributeTargets.Method, Inherited = true, AllowMultiple = false)]
	sealed class RunOnUIAttribute : Attribute, IWrapSetUpTearDown, IWrapTestMethod
	{
		public TestCommand Wrap(TestCommand command) => new RunOnUICommand(command);

		class RunOnUICommand : DelegatingTestCommand
		{
			public RunOnUICommand(TestCommand innerCommand)
				: base(innerCommand)
			{
			}

			public override TestResult Execute(TestExecutionContext context)
			{
				return Application.Instance.Invoke(() => innerCommand.Execute(context));
			}
		}
	}
```

Only downside (which is minor), is that you can't set this attribute on the test fixture to apply to all of its tests.

**Comment by jnm2** [comment]
Yes, great! I was going to suggest `IWrapSetUpTearDown` which you might prefer if you ever have a `[SetUp]` method etc.

> Only downside (which is minor), is that you can't set this attribute on the test fixture to apply to all of its tests.

Yes! See https://github.com/nunit/nunit/issues/2220.

Glad you figured it out! Feel free to ask further questions here.

Just curious: why not do the UI work in the test thread? What UI framework are you using that prevents you from using a test worker thread (STA or non-STA)?

**Comment by cwensley** [comment]
@jnm2 most UI frameworks are not multi-thread aware.  For example, macOS (Xamarin.Mac) and Gtk# as far as I can tell can't have any UI in a separate thread like WinForms or WPF can.  And to keep the main UI responsive (to show results, etc), the test engine needs to run in a background thread.

Many controls on macOS do actually need a message loop for them to run properly, so executing any UI on the test thread means there is no message loop running on that thread.

I am using this for an [NUnit UI](https://github.com/picoe/Eto/blob/develop/test/Eto.Test/Sections/UnitTestPanel.cs) written in [Eto.Forms](https://github.com/picoe/Eto), which abstracts WPF, WinForms, Xamarin.Mac, and GTK#.  I'm making a reusable version of the UI for apps built with Eto.Forms (and maybe a standalone UI).

Thanks for pointing to the other issue, I'll keep my eye on that one as well.

**Comment by CharliePoole** [comment]
@cwensley I haven't run your code, but looking at it I think your wrapper will get inserted into the command chain twice, once above the SetUpTearDown command and once below it. 

As described in the comments, the intention is that you should use __one__ of the interfaces derived from ICommandWrapper. Assuming your SetUp, Test and TearDown all need to run on the UI thread, you should probably be implementing IWrapSetUpTearDown only.

**Comment by CharliePoole** [comment]
@cwensley How are you finding ETO forms? I'm looking for a gui abstraction for use in the TestCentric GUI.

**Comment by cwensley** [comment]
@CharliePoole, ah you are right, thanks for pointing that out!

I am actually the author of Eto.Forms, so I guess I find it works really well if you want a UI that uses each platform's native UI toolkit. (;

**Comment by jnm2** [comment]
@cwensley 

It sounds like you're saying that the same UI thread and same message pump is shared between multiple tests. But what if you start the message pump inside each test case, on the test worker thread, and tear it down so that nothing is shared between tests? That way you don't need an extra UI thread.

Windows Forms and WPF can't share UI across threads either, due to Windows' native constraints. But, all that means is that you can't operate on the same window between tests.
As far as I have seen, this isn't really different from an OS perspective than if it was two separate processes opening two separate windows. I would be very surprised if macOS and Gtk# could not run two windows simultaneously with one message pump running on each thread, so long as they are completely isolated from each other. This is not something you would ever do in an application, but it is quite handy in tests. Not having to share state makes things simple and parallelization.

As someone who also develops native controls directly, thanks for humoring my curiosity! :D

**Comment by CharliePoole** [comment]
@cwensley Current plan is to do one release of the TestCentric GUI for NUnit using SWF and then settle on a different architecture for version 2.0. When I can find some more time, I'll hit you up offline for some advice.

**Comment by cwensley** [comment]
@jnm2,
> But what if you start the message pump inside each test case, on the test worker thread, and tear it down so that nothing is shared between tests? That way you don't need an extra UI thread.

That still wouldn't solve the problem.  The message pump would still have to be on a different thread because during the test code, the pump won't be "pumping". It'd still have to invoke into that separate thread. 

Another cool benefit having the tests invoke into the UI thread is the ability to show windows for "manual tests". I have a lot of tests that have to be verified visually by a human, which I show a form, provide instructions, and then the user clicks on pass or fail. This is very useful to verify certain behaviours that can't be programmatically tested. 

@CharliePoole, No problem, I'd be glad to help out if you decide to use Eto.Forms. I'll be updating my unit test UI as time goes on and you're more than welcome to use that as a starting point.

**Comment by cwensley** [comment]
@jnm2 btw, from [apple's docs](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/ThreadSafetySummary/ThreadSafetySummary.html):
> The following classes must be used only from the main thread of an application.
> - NSCell and all of its descendants
> - NSView and all of its descendants. For more information, see NSView Restrictions.

I tried creating a separate thread for the UI, and kaboom.

**Comment by jnm2** [comment]
> That still wouldn't solve the problem. The message pump would still have to be on a different thread because during the test code, the pump won't be "pumping". It'd still have to invoke into that separate thread.

Fyi, NUnit 3.11 will now be pumping messages for winforms and WPF as long as the test code is awaiting something. (https://github.com/nunit/nunit/issues/2123) Otherwise, you'd have to call Application.Run and Exit manually, which I also do in non-async UI tests.

> Another cool benefit having the tests invoke into the UI thread is the ability to show windows for "manual tests". I have a lot of tests that have to be verified visually by a human, which I show a form, provide instructions, and then the user clicks on pass or fail. This is very useful to verify certain behaviours that can't be programmatically tested.

I do this too, but directly on any test worker thread without spinning up a separate UI thread.

> The following classes must be used only from the main thread of an application.

This is pretty definitive. Looks like if I ever write macOS UI tests, I'll have to do exactly what you're doing. 👍

**Comment by mikkelbu** [comment]
@jnm2 You marked this with label "closed:done" and put it into the "closed without action" milestone, but the issue is still open?

**Comment by jnm2** [comment]
@mikkelbu Thanks! Meant to come back and close without discouraging conversation.

Existing issues: https://github.com/nunit/nunit/issues/2917, https://github.com/nunit/nunit/issues/2220.
